### PR TITLE
MXNet: add DistributedTrainer for Gluon

### PR DESCRIPTION
**Pull request by yuxihu**
DistributedTrainer is a subclass of Gluon Trainer. We override the `_allreduce_grads` function in Trainer to use Horovod `allreduce_` instead of kvstore pull/push methods. This makes distributed training using MXNet Gluon API with PS and Horovod have the same internal workflow.

Thanks @ptrendx and @romerojosh for proposing this [idea](https://github.com/horovod/horovod/pull/915#issuecomment-474022074).

### CODE REVIEW

**Comment by yuxihu** [comment]
@alsrgv @apeforest @eric-haibin-lin @ptrendx @romerojosh Please help review.

File: horovod/mxnet/__init__.py [hunk 1]
```diff
@@ -68,6 +68,17 @@ def set_wd_mult(self, args_wd_mult):
         self._optimizer.set_wd_mult(args_wd_mult)
 
 
+class DistributedTrainer(mx.gluon.Trainer):
+    def __init__(self, params, optimizer, optimizer_params=None):
+        super(DistributedTrainer, self).__init__(
```
Comment (romerojosh) [inline horovod/mxnet/__init__.py]: Do you think it would be a good idea to add something here to deal with the case where a user erroneously passes a `DistributedOptimizer` into the `DistributedTrainer`? You could just automatically unwrap the `DistributedOptimizer` for them (and maybe print a warning message).

Comment (yuxihu) [reply horovod/mxnet/__init__.py]: Make sense. Will add the check.

**Comment by yuxihu** [comment]
@alsrgv Looks like [Keras test](https://travis-ci.org/horovod/horovod/jobs/509631730#L9825) is failing.

File: README.md [hunk 2]
```diff
@@ -235,8 +232,10 @@ params = model.collect_params()
 if params is not None:
     hvd.broadcast_parameters(params, root_rank=0)
 
-# Create trainer and loss function
-trainer = gluon.Trainer(params, opt, kvstore=None)
+# Create DistributedTrainer, a subclass of gluon.Trainer
```
Comment (apeforest) [inline README.md]: Gluon was not mentioned in any context.

**Comment by alsrgv** [comment]
@yuxihu, the tests are broken due to https://github.com/tensorflow/tensorflow/issues/27067.  BTW, is it possible to switch all these https://github.com/apache/incubator-mxnet/blob/791c5d5e8091c8c16d1a30df1b0d53e4af60d245/include/mxnet/engine.h#L122 to function pointers?  Then we could avoid gcc4/gcc5 incompatibility, too.

File: horovod/mxnet/__init__.py [hunk 3]
```diff
@@ -68,6 +69,22 @@ def set_wd_mult(self, args_wd_mult):
         self._optimizer.set_wd_mult(args_wd_mult)
 
 
+class DistributedTrainer(mx.gluon.Trainer):
```
Comment (eric-haibin-lin) [inline horovod/mxnet/__init__.py]: Since this implements the Trainer API, are all APIs like save_state/load_state validated and tested?

**Comment by yuxihu** [comment]
@alsrgv I see. It looks like the change which broke the CI will not be reverted. What's the plan for Horovod CI?

It won't work by simply changing `std:function` to function poiner in the MXNet Engine. Lambda function with capture cannot be passed to a function pointer interface. There are more work to be done besides changing the signature which I am working with priority.

File: README.md (hunk 2 above)
Comment (yuxihu) [reply README.md]: Done

File: horovod/mxnet/__init__.py (hunk 3 above)
Comment (yuxihu) [reply horovod/mxnet/__init__.py]: I manually tested save_states and load_states and it worked. The DistributedTrainer is a subclass of Trainer which we only overrided one function `_allreduce_grads`.  This function has been covered by the MNIST example which we run in Horovod CI. For other APIs, the Python inheritance rules should govern and are covered by our MXNet unit tests.

**Comment by alsrgv** [comment]
@yuxihu, could you rebase this PR on latest master as well?

**Comment by alsrgv** [comment]
@apeforest @eric-haibin-lin @romerojosh, good to merge?

**Comment by romerojosh** [comment]
LGTM!

File: horovod/mxnet/__init__.py [hunk 4]
```diff
@@ -68,6 +69,22 @@ def set_wd_mult(self, args_wd_mult):
...
+    def __init__(self, params, optimizer, optimizer_params=None):
```
Comment (eric-haibin-lin) [inline horovod/mxnet/__init__.py]: It is worthwhile to document the difference between `DistributedTrainer` and `Trainer`, that `DistributedTrainer` performs allreduce(summation) and average, while `Trainer` only performs allreduce(summation)? That will help existing Trainer + kvstore users understand how gradient is calculated

Comment (yuxihu) [reply horovod/mxnet/__init__.py]: Added

**Comment by yuxihu** [comment]
@apeforest @eric-haibin-lin Ready for final review. Please take a look.
